projects/step-frontend/src/lib/modules/timeseries/metric-chart/metric-chart.component.ts

```diff
@@ -177,6 +219,45 @@ export class MetricChartComponent implements OnInit, OnChanges {
     };
   }
 
+  /**
+   * This function will fetch the entities found in the labels, and replace their ids with the appropriate name.
+   * @param series
+   * @param groupDimensionIndex
+   * @param fetchByIds
+   * @param mapEntityToLabel
+   * @private
+   */
+  private fetchAndSetLabels<T>(
+    series: TSChartSeries[],
+    groupDimensionIndex: number,
+    fetchByIds: (ids: string[]) => Observable<T[]>,
+    mapEntityToLabel: (entity: T) => string | undefined
+  ) {
+    const ids: string[] = series.map((s) => s.labelItems[groupDimensionIndex] as string).filter((id) => id);
+
+    if (ids.length > 0) {
+      fetchByIds(ids).subscribe((entities) => {
+        const entitiesByIds = new Map<string, T>();
+        entities.forEach((entity) => {
+          // Assuming each entity has an 'id' property
+          const entityId = (entity as any).id as string; // Cast to any to access 'id' property
+          if (entityId) {
+            entitiesByIds.set(entityId, entity);
+          }
+        });
+        series.forEach((s) => {
+          const entityId = s.labelItems[groupDimensionIndex];
+          if (entityId) {
+            const entity = entitiesByIds.get(entityId);
+            if (entity) {
+              this.chart.setLabelItem(s.id, groupDimensionIndex, mapEntityToLabel(entity));
+            }
+          }
+        });
+      });
```

Comment:
In case of Observables it make sense to do all required transformations inside the pipe chain,
and then do the final action inside the subscribe.

```suggestion
      fetchByIds(ids)
          .pipe(
              map((entities) => entities.reduce((res, entity) => {
                  const entityId = (entity as any).id as string;
                  if (entityId) {
                     res[entityId] = entity;
                  }
                  return res;
              }, {} as Record<string, T>)),
              map((entitiesDictionary) => series
                    .map(({id, labelItems}) => {
                         const entity = entitiesDictionary[labelItems[groupDimensionIndex]];
                         const entityLabel = entity ? mapEntityToLabel(entity) : undefined;
                         return {id, entityLabel};
                    })
                    .filter((seriesLabels) => !!seriesLabels.entityLabel)
              )
          )
          .subscribe((seriesLabels) => { 
               seriesLabels.forEach(({id, entityLabel}) => this.chart.setLabelItem(id, groupDimensionIndes, entityLabel);
          });
```

Reply:
@dvladir this looks to me as over-engineering . I see no reason to do it: it just adds complexity - both visually and technically. It's my code and I barely follow what the code is doing. If you consider it is a need though, I will add it.

Reply:
@lucianbaciu In concrete this case, both variants will work. I would say that this remark is not strictly required.
But in common, from the point of view how we work with observable streams, it make sense to do the transformations in pipe chains, to receive another stream with required data.
Then we can decide, will we subscribe to this new stream, will we reuse it as part of another stream or even will we pass the new stream as parameter somewhere else.
In other words wrapping data in the observable streams, increases the abstraction level

Reply:
@dvladir thank you. I know and I agree. If there would have been another other async task, or chains, I would definitely go for your suggestion.